proposals/custom-equals-in-value-classes.md

```diff
@@ -0,0 +1,262 @@
+star-projection of the value class.
+
+### Custom `hashCode`
+
+It is obvious that the ability of overriding `equals` makes us allow overriding of `hashCode`.
+
+### `@TypedEquals` and `@AllowTypedEquals` annotations
+
+Customization of equals in value classes is currently considered as an experimental feature. Thus, we want to force
+user to `@OptIn` before using it. In previous section we introduced `@TypedEquals` annotation, which serves as a marker
+for typed equals function. Unfortunately, simply annotating `@TypedEquals` with `@RequireOptIn` will not work, as
+typed equals may be invoked implicitly:
+
+```Kotlin
+@JvmInline
+value class IC(val x: Int) {
+    @TypedEquals
+    fun equals(other: IC) = ...
+}
+
+fun foo() {
+    setOf(IC(1), IC(2)) // compiles without @OptIn, but invokes typed equals internally
+}
+```
+
+We propose to add new annotation `@AllowTypedEquals`. This annotation is applicable for value classes only and any value
+class that declares typed equal must be annotated by `@AllowTypedEquals`:
+
+```Kotlin
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
+@MustBeDocumented
+@RequiresOptIn
+annotation class AllowTypedEquals
+```
+
+```Kotlin
+@JvmInline
+@AllowTypedEquals
+value class IC(val x: Int) {
+    @TypedEquals
+    fun equals(other: IC) = ...
+}
+
+@OptIn(AllowTypedEquals::class) // usage of value class IC requires OptIn
+fun foo() {
+    setOf(IC(1), IC(2))
+}
+```
+
+## Other questions
+
+### Inline classes over inline classes
+
+Suppose we have an inline class `B` over inline class `A` and `A` declares typed equals. In this case,
+default-generated equals in `B` will be comparing it underlying values using custom equals of `A`.
+
+### Interaction with Valhalla
+
+In its current prototype version `20-valhalla+20-75` Valhalla supports customization of equals and hashCode in values
+classes, thus, we will
+be able to compile Kotlin value classes with custom equals to Valhalla values classes.
+
+```Java
+value class ValueClass {
+    int x;
+
+    public ValueClass(int x) {
+        this.x = x;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        return 42;
+    }
+}
+
+class Main {
+    public static void main(String[] args) {
+        ValueClass zeroVal = new ValueClass(0);
+        ValueClass oneVal = new ValueClass(1);
+
+        Object zeroObj = zeroVal;
+
+        System.out.println(zeroVal.equals(oneVal));        // true
+        System.out.println(zeroObj.equals(oneVal));        // true
+        System.out.println(zeroObj.hashCode());            // 42
+    }
+}
+```
+
+## Possible future improvements
+
+### Resolving references to typed equals in compiler frontend
+
+Currently `'=='` operator is being resolved to typed equals in compiler backend, but resolving it in frontend could let
```

Comment:
Current decision is that `==` is never resolved on front-end, but IDE makes everything it can to resolve it. So, it is not about FE support, but rather IDE one.